subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+[source,kotlin]
+----
+// Without sam-with-receiver
+tasks.register("hello") { task ->
+    // You need to use the lambda parameter
+    task.doLast {
+        println("Hello World")
+    }
+}
+
+// With sam-with-receiver
+tasks.register("hello") {
+    // You can use the receiver
+    doLast {
+        println("Hello World")
+    }
+}
+----
+
+The latter looks closer to `build.gradle.kts` scripts and more concise but is more prone to scoping errors as well.
+
+
+[[sec:kotlin_assignment]]
+=== `kotlin-assignment` compiler plugin
+
+The `kotlin-assignment` compiler plugin is still experimental. It allows overriding assignments so that Kotlin callers do not have to call `.set()` on https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Property.html[Property]:
+
+
+----
+// Without kotlin-asignment
+tasks.jar.configure {
+    // You need to call .set()
+    archiveBaseName.set("mylib")
+}
+
+// With kotlin-asignment
+tasks.jar.configure {
+    // No need to call .set()
+    archiveBaseName = "mylib"
+}
+----
+
+[[sec:api_version]]
+=== `apiVersion` compiler option
+
+The https://kotlinlang.org/docs/compiler-reference.html#api-version-version[Kotlin compiler apiVersion option] enforces that the generated bytecode only uses declaration from the given version. This is important to stay compatible with the Kotlin version embedded by Gradle.
+
+Note: while this enforces your code is compatible with a given version of Kotlin stdlib and Gradle, it does not check dependencies, so it is important to check your dependencies are also compatible.
+
+[[sec:sam_conversion_class]]
+=== `-Xsam-conversions=class` compiler option
+
+`-Xsam-conversions=class` sets up the implementation strategy for SAM (single abstract method) conversion to always generate anonymous classes, instead of using the `invokedynamic` JVM instruction. This is to make sure the lambdas are serializable and provide a better support for configuration cache and incremental build. This is particularly important for lambdas used from `doLast:
+See https://github.com/gradle/gradle/issues/17052[Issue #17052] for more information.
+
+[source,kotlin]
+----
+// Without `-Xsam-conversions=class` task will never be up-to-date
+task.doLast {
+    // Something
+}
+----
+
+
+[[sec:plugins_for_your_plugins]]
+== Different development plugins
+
+In order to configure all the above options and more, Gradle offers several plugins:
+
+* `"java-gradle-plugin"`
+* `kotlin("jvm")` (id `"org.jetbrains.kotlin.jvm"`)
+* `embedded-kotlin` (id `"org.gradle.kotlin.embedded-kotlin"`))
+* `kotlin-dsl` (id `"org.gradle.kotlin.kotlin-dsl"`)
```

Comment:
🔍 I had a discussion with the Gradle team regarding the messy way to apply plugins in Kotlin; it's worth mentioning how confusing it can be to users: accessor (sometimes with backtick) vs `kotlin(<identifier>)` vs `id(<plugin-id>)`. The general conclusion was to prefer `id(<plugin-id>)` but it cannot work in all cases. We may want to take that into consideration here.

Reply:
+1. Having 3 ways to specify a plugin id is adding entropy in the system.

I'm team `id(<plugin-id>)` all the time. `embedded-kotlin` and `kotlin-dsl` are special because their version (but not the their implementation) is embedded into the Gradle distribution. My favourite solution there would be to have the Gradle core know about those ids and fail if a version is specified:

```
plugins {
  // Gradle knows about this and pulls the compatible version automatically
  id("org.gradle.kotlin.kotlin-dsl")
}

plugins {
  // Error: You cannot specify the version of org.gradle.kotlin.kotlin-dsl
  id("org.gradle.kotlin.kotlin-dsl").version("4.0.6")
}

plugins {
  // Users could still use a different "unsafe" pluginId if they know what they're doing
  // (naming up for grabs)
  id("org.gradle.kotlin.kotlin-dsl.unsafe").version("4.0.6")
}
```

Also I'd personally remove `kotlin("jvm")`. My version catalog files are already full of maven coordinates that I can't remember so adding one pluginId for Kotlin doesn't hurt and reduces the cognitive load of seeing `kotlin("jvm")` vs `org.jetbrains.kotlin.jvm`.

Reply:
I completely agree.